src/Storages/MergeTree/MergeTask.h

```diff
@@ -0,0 +1,253 @@
+#pragma once
+
+#include <Storages/MergeTree/IExecutableTask.h>
+#include <Storages/MergeTree/MergeProgress.h>
+#include <Storages/MergeTree/MergeTreeData.h>
+#include <Storages/MergeTree/IMergedBlockOutputStream.h>
+#include <Storages/MergeTree/MergedBlockOutputStream.h>
+#include <Storages/MergeTree/FutureMergedMutatedPart.h>
+#include <Storages/MergeTree/ColumnSizeEstimator.h>
+#include <Storages/MergeTree/MergedColumnOnlyOutputStream.h>
+#include <DataStreams/ColumnGathererStream.h>
+#include <Compression/CompressedReadBufferFromFile.h>
+
+#include <memory>
+#include <list>
+
+namespace DB
+{
+
+class MergeTask;
+using MergeTaskPtr = std::shared_ptr<MergeTask>;
+
+/**
+ * Overview of the merge algorithm
+ *
+ * Each merge is executed sequentially block by block.
+ * The main idea is to make a merge not a subroutine which is executed
+ * in a thread pool and may occupy a thread for a period of time,
+ * but to make a merge a coroutine which can suspend the execution
+ * in some points and then resume the execution from this point.
+ *
+ * A perfect point where to suspend the execution is after the work over a block is finished.
+ * The task itself will be executed via BackgroundJobExecutor.
+ *
+ * The interface of the task is simple.
+ * The main method is `execute()` which will return true, if the task wants to be executed again and false otherwise.
+ *
+ * With this kind of task we can give a merge a priority.
+ * A priority is simple - the lower the size of the merge, the higher priority.
+ * So, if ClickHouse wants to merge some really big parts into a bigger part,
+ * then it will be executed for a long time, because the result of the merge is not really needed immediately.
+ * It is better to merge small parts as soon as possible.
+*/
+class MergeTask
```

Comment:
Well, we should find some way how to improve this code.
It is mainly only moved code, but now it became a little bit unreadable.
We can kind of leave it that way, but it will be a tech depth. I am ok if we have another one following pr with it.

A couple of ideas what to do:
* constructor has a lot of params - maybe group them into one or several structs, also add comment to them
* there are several states and according functions, but prev variables from stack are now all together. maybe we could create separate struct for each stage, which is alive only while stage is being executed
* also maybe replace enum with stages to an array of member functions + current index - because we execute them in order.

Reply:
BTW MutateTask looks more readable